[WIP] Update Manager Refactor

jaxwilko opened the pull request:
Me and Luke are working on a little refactor of how the UpdateManager works...

It also includes some components from my old composer PR found here: https://github.com/wintercms/winter/pull/967

The core idea of this is that there should be consistent interfaces to interact with different things built on top of Winter (i.e. `modules`, `plugins` & `themes`). Each of these elements will now have their class (i.e. `Theme`, `PluginBase`, `ModuleServiceProvider`) will now extend `WinterExtension` which will allow them to all interact with the update manager and other system functions directly.

In addition `PluginManager`, `ThemeManager` & `ModuleManager` (new) will all implement the `ExtensionManager` interface, which will allow for a consistant way to interact with all Winter extensions.

I.e. You will be able to call `PluginManager::instance()->update($plugin)` or `ThemeManager::instance()->update($theme)`.

This also allows us to do some cool stuff like (only support for plugin installing is currently functional)
```php
(new ExtensionSource(
    ExtensionSource::SOURCE_COMPOSER,
    ExtensionSource::TYPE_PLUGIN,
    composerPackage: 'jaxwilko/datamigrator'
))->install();

(new ExtensionSource(
    ExtensionSource::SOURCE_COMPOSER,
    ExtensionSource::TYPE_THEME,
    composerPackage: 'jaxwilko/custom-theme'
))->install();

(new ExtensionSource(
    ExtensionSource::SOURCE_LOCAL,
    ExtensionSource::TYPE_PLUGIN,
    code: 'winter.demo'
))->install();

(new ExtensionSource(
    ExtensionSource::SOURCE_LOCAL,
    ExtensionSource::TYPE_PLUGIN,
    path: 'plugins/winter/demo'
))->install();
```

By the end of the PR, the UpdateManager may no longer exist as is, instead each `ExtensionManager` will be responsible for its existing functionality.

Requires storm branch: https://github.com/wintercms/storm/tree/wip/support-uploading-packages-update-manager-changes

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added marketplace UI for browsing and installing plugins and themes directly.
  * Added support for uploading plugins and themes as ZIP files.
  * Integrated marketplace API for enhanced package management and discovery.

* **Improvements**
  * Redesigned plugin and theme management interface with improved navigation and search capabilities.
  * Updated system updates page with streamlined extension installation workflow.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

LukeTowers committed:
Added initial type hinting and better imports for update managing classes

LukeTowers committed:
Fix tests

LukeTowers committed:
add missing newline

LukeTowers committed:
Initial implementation of the UI to upload plugins via zip folder

LukeTowers committed:
Centralize updatemanager zip extracting

LukeTowers committed:
Add tests & fixtures for initial plugin scanning functionality

LukeTowers committed:
Initial work on supporting .zip files from plugin:install command.

LukeTowers committed:
formatting

LukeTowers committed:
Update plugin test fixtures to avoid conflicts when installing them

LukeTowers committed:
Initial working install method for plugin:install command

LukeTowers committed:
Hopefully fix windows tests

LukeTowers committed:
WIP for retrieving the uploaded file from the plugin uploader modal

LukeTowers committed:
Merge branch 'develop' into wip/support-uploading-packages

LukeTowers committed:
Merge branch 'develop' into wip/support-uploading-packages

LukeTowers committed:
Merge branch 'develop' into wip/support-uploading-packages

jaxwilko committed:
Init wip

jaxwilko committed:
Work in progress... do not try and run it :(

LukeTowers committed:
Add theme upload form config

jaxwilko committed:
Added winter/packager + ovberrides and bumped php version

jaxwilko committed:
Added wip

jaxwilko committed:
Added fix to allow install to create files

jaxwilko committed:
Added objects for source types

jaxwilko committed:
WIP

jaxwilko committed:
Added wip

jaxwilko committed:
Added fix for installing specific package

jaxwilko committed:
Added abstract stubs

jaxwilko committed:
Added support for updating composer plugins

jaxwilko committed:
Added wip

jaxwilko committed:
Migrated core code to storm

jaxwilko committed:
Added wip

jaxwilko committed:
Added wip

jaxwilko committed:
Moved commands into directories

jaxwilko committed:
Refactored stuff

jaxwilko committed:
Added theme extension logic

jaxwilko committed:
Added market place support

jaxwilko committed:
Added plugin code detection from path using child process

jaxwilko committed:
Added support for installing from arbitary paths

jaxwilko committed:
Added theme successfully installed message

jaxwilko committed:
Added plugin successfully installed message and fixed no updates required causing an error

jaxwilko committed:
Added more messages to improve ux

jaxwilko committed:
Added wip

jaxwilko committed:
Hooked up old update checker ui with new update system

LukeTowers commented on composer.json:
```diff
@@ -29,13 +29,14 @@
     },
     "require": {
-        "php": "^8.0.2",
+        "php": "^8.1",
```
What's the PHP 8.1 feature that we're using?

jaxwilko replied:
Interface const overriding (see: https://www.php.net/manual/en/language.oop5.interfaces.php), but we're not using it any more so could rollback that change

LukeTowers committed:
Update composer.json

jaxwilko committed:
Moved stubs for create commands

jaxwilko committed:
Added better reporting for module upgrades

jaxwilko committed:
Removed stubbed function as implemented in parent class

jaxwilko committed:
Added default return

jaxwilko committed:
Added fix to ensure upgrades succeed for uninstalled plugins

jaxwilko committed:
Added fix for reporting

jaxwilko committed:
Added support for preserving modules

jaxwilko committed:
Removed call to setOutputNotes

jaxwilko committed:
Added fixes for tests

jaxwilko committed:
Merge branch 'wip/support-uploading-packages-update-manager-changes' of github.com:wintercms/winter into wip/support-uploading-packages-update-manager-changes

jaxwilko committed:
Removed packager from winter deps

jaxwilko committed:
Added style fixes

jaxwilko committed:
Reverted php requirement change

jaxwilko committed:
Added findThemesInPath method

jaxwilko committed:
Added method for unpacking a zip and returning all sources inside

LukeTowers committed:
Fix typo

LukeTowers commented on modules/backend/ServiceProvider.php:
```diff
@@ -322,4 +323,14 @@ protected function registerBackendSettings()
+
+    public function getPath(): string
+    {
+        return __DIR__;
```
Could probably move this into the base ModuleServiceProvider with the following:

```php
public function getPath() {
        $reflector = new \ReflectionClass(get_called_class());
        return dirname($reflector->getFileName());
    }
```

jaxwilko replied:
See https://github.com/wintercms/winter/pull/1259/commits/729e46e0abcffdc86cc3609314156dad408497aa & https://github.com/wintercms/storm/commit/6b03b16ff901b953db6322930aab4859b1d53235

LukeTowers commented on modules/backend/ServiceProvider.php:
```diff
@@ -322,4 +323,14 @@ protected function registerBackendSettings()
+
+    public function getIdentifier(): string
+    {
+        return 'Backend';
```
Could probably move this into the base ModuleServiceProvider with the following:

```php
public function getIdentifier(): string
{
    $reflector = new \ReflectionClass(get_called_class());
    return $reflector->getNamespaceName();
}
```

jaxwilko replied:
See https://github.com/wintercms/winter/pull/1259/commits/729e46e0abcffdc86cc3609314156dad408497aa & https://github.com/wintercms/storm/commit/6b03b16ff901b953db6322930aab4859b1d53235

LukeTowers commented on modules/cms/ServiceProvider.php:
```diff
@@ -477,4 +491,14 @@ protected function registerHalcyonModels()
+    public function getIdentifier(): string
+    {
+        return 'Cms';
+    }
```
Same as Backend ServiceProvider comments

jaxwilko replied:
See https://github.com/wintercms/winter/pull/1259/commits/729e46e0abcffdc86cc3609314156dad408497aa & https://github.com/wintercms/storm/commit/6b03b16ff901b953db6322930aab4859b1d53235

LukeTowers commented on modules/system/classes/extensions/ExtensionManager.php:
```diff
@@ -0,0 +1,104 @@
+        if ($container->bound(static::class)) {
+            $container->forgetInstance(static::class);
+        }
+    }
```
Why not use the Singleton trait?

jaxwilko replied:
Singleton disallows `__constructor()` which we use to define the default output object... could work round it tho by checking if none exists and creating at use
![image](https://github.com/user-attachments/assets/97aa9f98-4133-4a3c-bfe3-0feb0a131383)

LukeTowers replied:
Sounds good, let's give that a try then.

LukeTowers commented on modules/system/classes/extensions/ExtensionManagerInterface.php:
```diff
@@ -0,0 +1,131 @@
+     *
+     * @return static
+     */
+    public function tearDown(): static;
```
Seems dangerous 😂 What's it used for?

jaxwilko replied:
it's uninstall all, `uninstall` only accepts a single extension, i.e. `PluginManager::instance()->uninstall('Jax.Test')`. I wanted to make it a seperate method to make it veryy obvious that what's about to happen.

I also didn't want a situation like:
```php
$toUninstall = ['A.Test', null, 'B.Test'];
foreach ($toUninstall as $plugin) {
    PluginManager::instance()->uninstall($plugin);
}
```
Oh no everything just got deleted, including all installed plugins not listed.

This way, it'll just throw a plugin not found exception and if you want to remove everything you have to explicitly call `tearDown`.

LukeTowers commented on modules/system/classes/extensions/ModuleManager.php:
```diff
@@ -0,0 +1,283 @@
+
+    public function create(string $extension): WinterExtension
+    {
+        throw new ApplicationException('Support for creating extensions needs implementing');
```
```suggestion
        throw new ApplicationException('Support for creating modules is not implemented.');
```

LukeTowers commented on modules/system/classes/UpdateManager.php:
```diff
@@ -39,178 +31,44 @@ class UpdateManager
     {
-        $this->migrator = App::make('migrator');
-        $this->repository = App::make('migration.repository');
+        return !Schema::hasTable($this->getMigrationTableName());
```
```suggestion
        return Schema::hasTable($this->getMigrationTableName());
```

LukeTowers commented on modules/system/classes/extensions/ModuleManager.php:
```diff
@@ -0,0 +1,283 @@
+    {
+        $modules = $this->getModuleList($extension);
+
+        $firstUp = UpdateManager::instance()->isSystemSetup();
```
```suggestion
        $firstUp = !UpdateManager::instance()->isSystemSetup();
```

LukeTowers commented on modules/system/classes/extensions/ModuleManager.php:
```diff
@@ -0,0 +1,283 @@
+            $this->output->info('Migration table created');
+        }
+
+        if (!$migrationsOnly && !Config::get('cms.disableCoreUpdates')) {
```
We should probably include an info or warning message if `cms.disableCoreUpdates` is set; I used to use it on every site to prevent the backend from updating the core when I was actually managing them with composer so I could see it causing issues if we silently ignore them using that setting.

jaxwilko replied:
see: https://github.com/wintercms/winter/pull/1259/commits/4e339308a4fbce24803adee44c9fbb84018c9146

LukeTowers commented on modules/system/classes/extensions/ModuleManager.php:
```diff
@@ -0,0 +1,283 @@
+                    ));
+
+                    Preserver::instance()->store($extension);
+                    $update = Composer::update(dryRun: true, package: $composerPackage);
```
Is the dryrun here a temporary thing?

jaxwilko replied:
Yes, I did leave a todo in the plugin manager for it. I may have deleted all my work but not including it so till we're ready I'm keen on leaving it in lol

LukeTowers commented on modules/system/classes/extensions/ModuleManager.php:
```diff
@@ -0,0 +1,283 @@
+                        $composerPackage
+                    ));
+
+                    Preserver::instance()->store($extension);
```
Is the intention to use this to restore the previous version if something fails, or perhaps to allow users to rollback to a specific version when it's been preserved?

jaxwilko replied:
Yes, it will keep a local archieve of the extension prior to update, this should be the groundwork for us to revert properly back to a previous version

LukeTowers commented on modules/system/classes/extensions/ModuleManager.php:
```diff
@@ -0,0 +1,283 @@
+            $this->output->info(sprintf('Migrating %s module...', $module));
+            $this->migrator->run(base_path() . '/modules/' . strtolower($module) . '/database/migrations');
+
+            if ($firstUp) {
+                $className = '\\' . $module . '\Database\Seeds\DatabaseSeeder';
+                if (class_exists($className)) {
+                    $this->output->info(sprintf('Seeding %s module...', $module));
```
Probably out of scope for this PR, but should we approach seeders (both module and plugin, which aren't quite the same) as needing the same level of tracking as migrations so that any that haven't been run yet will get run? With the current code, if a new module is added after the project is setup then it will never have its seeder run because it'll never be the firstUp again.

jaxwilko replied:
Agreed, maybe we could use `Parameter` to record inits for modules? It's currently used like that for themes.

LukeTowers replied:
Works for me

LukeTowers commented on modules/system/classes/extensions/ModuleManager.php:
```diff
@@ -0,0 +1,283 @@
+
+        if (!$preserveFiles) {
+            // Modules probably should not be removed
+            // File::deleteDirectory($module->getPath());
```
When would this method be called and the files not need to be removed?

LukeTowers replied:
Also something to think about with the uninstall() method is removal from the composer.json

jaxwilko replied:
When you run `./artisan winter:down && ./artisan winter:up` and he won't come up because you just deleted system :'(

LukeTowers replied:
Right, but shouldn't those commands be passing the preserve files flag?

LukeTowers commented on modules/system/classes/extensions/PluginManager.php:
```diff
@@ -120,8 +120,389 @@ protected function init(): void
+    }
+
+    /**
+     * @throws SystemException
```
What throws the SystemException?

jaxwilko replied:
I think it was something from a previous revision, see: https://github.com/wintercms/winter/pull/1259

LukeTowers commented on modules/system/classes/extensions/PluginManager.php:
```diff
@@ -120,8 +120,389 @@ protected function init(): void
+                throw new ApplicationException(Lang::get('system::lang.plugins.replace.multi_install_error'));
+            }
+            foreach ($replaces as $replace) {
+                $this->versionManager()->replacePlugin($plugin, $replace);
```
Why is there a method for accessing the property?

jaxwilko replied:
Public method, private property. Property should be used internally tho, this code was moved from `UpdateManager` and this refactor was overlooked. See: https://github.com/wintercms/winter/pull/1259/commits/c8bbe35b90f157ca5f4e832bf29396bf5d8d177e

LukeTowers commented on modules/system/classes/extensions/PluginManager.php:
```diff
@@ -1034,41 +1421,83 @@ protected function sortByDependencies(): array
-    //
-    // Management
-    //
+    public function resolveIdentifier(ExtensionSource|WinterExtension|string $extension): ?string
```
Should we be more specific in the manager resolveIdentifier() methods to only accept the specific WinterExtension implementations that we accept?

jaxwilko replied:
It's overriding the parent, not sure if we can limit the scope of the type hints

LukeTowers commented on modules/system/classes/extensions/PluginVersionManager.php:
```diff
@@ -17,25 +21,18 @@
  * @author Alexey Bobkov, Samuel Georges
  */
-class VersionManager
+class PluginVersionManager
```
Is the only reason this is a separate class to avoid having the PluginManager being too long or is it ever used by anything else?

jaxwilko replied:
It is used by some other parts of the system but only via the plugin manager, i.e. `PluginManager::instance()->versionManager()`. It could probably be merged in but `PluginManager` is already chonk and confusing...

LukeTowers commented on modules/system/classes/extensions/source/ExtensionSource.php:
```diff
@@ -0,0 +1,354 @@
+
+        // Create a full path to the plugin.php file
+        $file = $path . DIRECTORY_SEPARATOR . $files['plugin.php'];
+
+        // The following will create a child process to parse the plugins namespace, this is done to prevent pollution
+        // of the current process (i.e. namespaces being incorrectly registered to paths where they won't be in the
+        // future post installation)
+
+        // Create a named pipe
+        $pipeName = 'classTestPipe';
+        posix_mkfifo($pipeName, 0777);
+
+        // Fork the process
+        if (!pcntl_fork()) {
+            // Open the pipe now, so we can signal failure by closing it in case of a parse error
+            $pipe = fopen($pipeName, 'w');
+            try {
+                require $file;
+            } catch (\Throwable $e) {
+                // Probably a parse error, close the pipe to tell the parent that something went wrong
+                fclose($pipe);
+                exit(1);
+            }
+            // Get all declared classes
+            $classes = get_declared_classes();
+            // Reflect on the last registered class, i.e. the one we loaded with require above
+            $reflect = new \ReflectionClass(array_pop($classes));
+            // Send the namespace over the pipe to the parent process
+            fwrite($pipe, $reflect->getNamespaceName());
+            fclose($pipe);
+            exit(0);
+        }
+
+        // Read the pipe, this is blocking and will wait for the child to close the pipe handle in it's own process
+        $pipe = fopen($pipeName, 'r');
+        // Read the namespace from the pipe
+        $namespace = fread($pipe, 4096);
+        // Close and clean up
+        fclose($pipe);
+        unlink($pipeName);
```
Crazy wizard 🧙 😂 
Let's just use regex here instead to scrape the namespace from the plugin.php. Also I think the only valid Plugin.php is `Plugin.php` verbatim, so probably no need to normalize the paths?

jaxwilko replied:
Not sure, we already do search for it in 2 different ways :shrug: 
https://github.com/wintercms/winter/blob/c8bbe35b90f157ca5f4e832bf29396bf5d8d177e/modules/system/classes/extensions/PluginManager.php#L883
https://github.com/wintercms/winter/blob/c8bbe35b90f157ca5f4e832bf29396bf5d8d177e/modules/system/classes/extensions/PluginManager.php#L920

jaxwilko replied:
See: https://github.com/wintercms/winter/pull/1259/commits/8da2212fc295e98b7f4aa5e12ecb4a3c29940895

LukeTowers commented on modules/system/console/BaseScaffoldCommand.php:
```diff
@@ -89,18 +90,27 @@ public function makeStubs(): void
             . 'lang.php'
         );
+
+        $relativeFile = str_replace(base_path(), '', $langFilePath);
```
Me likey ❤️

LukeTowers commented on modules/system/console/JaxTest.php:
```diff
@@ -0,0 +1,49 @@
+
+        return 0;
+    }
+}
```
Does this file still need to be here?

jaxwilko replied:
It's my test, will remove before merge, just commit so it stopped being annoying

LukeTowers commented on modules/system/controllers/updates/traits/ManagesMarketplaceProject.php:
```diff
@@ -0,0 +1,61 @@
+use Exception;
+use Illuminate\Http\RedirectResponse;
+use Illuminate\Support\Facades\Lang;
+use October\Rain\Exception\ApplicationException;
```
```suggestion
use Winter\Storm\Exception\ApplicationException;
```

LukeTowers commented on modules/system/tests/README.md:
````diff
@@ -59,8 +59,7 @@ php artisan plugin:refresh Acme.Blog
 ```php
-use System\Classes\PluginManager;
-use System\Tests\Bootstrap\PluginTestCase;
+use System\Classes\Extensions\PluginManager;use System\Tests\Bootstrap\PluginTestCase;
````
```suggestion
use System\Classes\Extensions\PluginManager;
use System\Tests\Bootstrap\PluginTestCase;
```

LukeTowers commented on modules/system/tests/bootstrap/PluginManagerTestCase.php:
```diff
@@ -52,7 +51,7 @@ public function setUp() : void
 
         // Forces plugin migrations to be run again on every test
-        VersionManager::forgetInstance();
+//        PluginVersionManager::forgetInstance();
```
Is this temporary?

jaxwilko replied:
Yes, it's a note to self that I need to look into how not flushing the instance works inside tests

LukeTowers commented on modules/system/traits/InteractsWithZip.php:
```diff
@@ -0,0 +1,33 @@
+            throw new ApplicationException(Lang::get('system::lang.zip.extract_failed', ['file' => $archive]));
+        }
+
+        @unlink($archive);
```
Is it expected that the original archive is always removed?

jaxwilko replied:
I would expect so? Reckon it should be an optional param?

LukeTowers replied:
Yes probably. If we're using the preserver then we wouldn't want to delete the copy after restoring it I would think

LukeTowers commented on modules/system/classes/UpdateManagerOld.php:
```diff
@@ -0,0 +1,1100 @@
+<?php
+?>
```
To be removed

jaxwilko replied:
See: https://github.com/wintercms/winter/pull/1259/commits/f950ce3b8e735ed574fe23e97bae5e9b7bcbb2e7

LukeTowers commented on modules/system/classes/UpdateManager.php:
```diff
@@ -5,27 +5,19 @@
-use Illuminate\Support\Facades\Lang;
-use RecursiveDirectoryIterator;
-use RecursiveIteratorIterator;
+use Mix\TestA\Plugin;
```
Was the wrong class imported here?

jaxwilko replied:
Yes, see: https://github.com/wintercms/winter/pull/1259/commits/16a4841c4c9bccd216c1da9770a34b6f4e9d6fc0

LukeTowers commented on modules/system/classes/InstallManager.php:
```diff
@@ -0,0 +1,8 @@
+
+namespace System\Classes;
+
+class InstallManager
```
What's the plan for this class?

jaxwilko replied:
See: https://github.com/wintercms/winter/pull/1259/commits/f9773db6ed248380fd1ae135395025214128ff34.

I think the original idea was to do basically what `ExtensionSource` is doing

LukeTowers reviewed (changes requested):
Overall very happy with the direction of this PR, great work so far @jaxwilko! ❤️

jaxwilko committed:
Moved logic into parent

jaxwilko committed:
Merge branch 'wip/support-uploading-packages-update-manager-changes' of github.com:wintercms/winter into wip/support-uploading-packages-update-manager-changes

mjauvin commented on modules/system/classes/extensions/source/ExtensionSource.php:
```diff
@@ -0,0 +1,354 @@
+    /**
+     * @throws ApplicationException
+     */
+    public function createFiles(): ?static
```
Why not move this logic to its dedicated ExtensionSource classes (i.e. ComposerSource, LocalSource, MaketSource)

jaxwilko replied:
Honestly because I only added those classes later after already implementing all of this logic in `ExtensionSource`, but yes I think it should be moved. Was just easier during development to keep them all in the same place.

jaxwilko replied:
Ah, so we might not want to do this as the internal `$source` property is changed during install, i.e. after `composer require ...` is executed and we have the files, we modify the `$source` to local so we don't re-download if `createFiles()` is called again. Just did the work to split it all out and not sure if I like it :thinking:

mjauvin replied:
Ah, makes sense to keep it in here then!

jaxwilko committed:
Apply suggestions from code review

Co-authored-by: Luke Towers <luke@luketowers.ca>

jaxwilko committed:
Added notice when cms.disableCoreUpdates is enabled

jaxwilko committed:
Removed misleading comment

jaxwilko committed:
Added fix to refer to property instead of method

jaxwilko committed:
Added use of plugin manager to check for plugin code

jaxwilko committed:
Removed old update manager

jaxwilko committed:
Removed incorrect imports

jaxwilko committed:
Removed unused class

jaxwilko committed:
Made plugin guessing even simplier

mjauvin commented on modules/system/classes/extensions/ModuleManager.php:
```diff
@@ -0,0 +1,286 @@
+
+                    $versions
+                        ? $this->renderComponent(Info::class, sprintf('Updated module %s (%s) from v%s => v%s', $module, $composerPackage, $versions[0], $versions[1]))
+                        : $this->renderComponent(Error::class, sprintf('Failed to module %s (%s)', $module, $composerPackage));
```
```suggestion
                        : $this->renderComponent(Error::class, sprintf('Failed to update module %s (%s)', $module, $composerPackage));
```

mjauvin commented on modules/system/classes/extensions/PluginManager.php:
```diff
@@ -1034,41 +1430,83 @@ protected function sortByDependencies(): array
+     * Disables the provided plugin using the provided flag (defaults to static::DISABLED_BY_USER)
+     * @deprecated
+     */
+    public function disablePlugin(PluginBase|string $plugin, string|bool $flag = self::DISABLED_BY_USER): bool
```
```suggestion
    public function disablePlugin(PluginBase|string $plugin, string|bool $flag = self::DISABLED_BY_USER): ?bool
```

mjauvin commented on modules/system/classes/extensions/PluginManager.php:
```diff
@@ -1034,41 +1430,83 @@ protected function sortByDependencies(): array
+     * Enables the provided plugin using the provided flag (defaults to static::DISABLED_BY_USER)
+     * @deprecated
+     */
+    public function enablePlugin(PluginBase|string $plugin, $flag = self::DISABLED_BY_USER): bool
```
```suggestion
    public function enablePlugin(PluginBase|string $plugin, $flag = self::DISABLED_BY_USER): ?bool
```

mjauvin reviewed (commented):
called method might return null, we need to use `?bool` return type hint

jaxwilko committed:
Apply suggestions from code review

Co-authored-by: Marc Jauvin <marc.jauvin@gmail.com>

mjauvin commented:
@jaxwilko @LukeTowers should we add localization strings afterward or in this PR for info/warning/error messages ?

mjauvin commented:
@jaxwilko `requestUpdateList()` does not exist anymore in UpdateManager class but is still being called in `check()` method.

mjauvin commented:
Also getting this error when clicking "install themes" from the `Updates & Plugins` page:
![image](https://github.com/user-attachments/assets/330894bd-cd96-4175-a279-883c3180f500)

jaxwilko commented:
> @jaxwilko @LukeTowers should we add localization strings afterward or in this PR for info/warning/error messages ?

100%

All the strings in output atm are just placeholders

jaxwilko commented:
> @jaxwilko `requestUpdateList()` does not exist anymore in UpdateManager class but is still being called in `check()` method.

Yeah, I've not finished fully migrating the backend yet, check for updates should work (but only supports composer packages) and interacting with plugins should work

jaxwilko commented:
> Also getting this error when clicking "install themes" from the `Updates & Plugins` page:
> ![image](https://github.com/user-attachments/assets/330894bd-cd96-4175-a279-883c3180f500)
> 

Luke broke it, it's from his PR this one is based off... I'll look into it when i do more on the backend stuff :)

jaxwilko committed:
Wip

jaxwilko commented:
@bennothommo plz need sexy help!
![image](https://github.com/user-attachments/assets/d1fc529c-3b6d-481e-a45d-d61e13d8d714)

LukeTowers commented on modules/system/controllers/updates/assets/src/components/Product.vue:
```diff
@@ -0,0 +1,151 @@
+    margin-right: 7px;
+}
+</style>
+
+
```
```suggestion
```

LukeTowers committed:
Update modules/system/controllers/updates/assets/src/components/Product.vue

LukeTowers commented on modules/system/controllers/updates/_install_plugins.php:
```diff
@@ -1,119 +1,8 @@
-        </a>
-    </div>
-</script>
+
```
```suggestion
```

LukeTowers committed:
Update modules/system/controllers/updates/_install_plugins.php

LukeTowers commented on modules/system/controllers/updates/assets/src/components/Product.vue:
```diff
@@ -0,0 +1,149 @@
+                <div title="Downloads" class="downloads">
+                    <span class="product-badge"><i class="icon-download"></i></span>
+                    {{product.downloads}}
+                </div>
```
We should probably add the license as well

jaxwilko replied:
Sure, it'll have to come from the marketplace tho as the packagist api does not provide it

jaxwilko committed:
Moved eslint to prevent compiled js being linted

jaxwilko committed:
Merge branch 'wip/support-uploading-packages-update-manager-changes' of github.com:wintercms/winter into wip/support-uploading-packages-update-manager-changes

jaxwilko committed:
Added fix for eslint ignore

LukeTowers commented:
For future reference, @jaxwilko's JS for the event stream handling

```js
const response = await fetch(window.location.href, {
                method: 'POST',
                headers: {
                    'Accept': 'text/event-stream',
                    'X-Requested-With': 'XMLHttpRequest',
                    'X-WINTER-REQUEST-HANDLER': 'onInstallPlugin',
                    'X-WINTER-REQUEST-PARTIALS': '',
                    'X-CSRF-TOKEN': document.querySelector('meta[name="csrf-token"]').getAttribute('content')
                },
                body: data,
                redirect: 'follow',
                mode: 'same-origin',
            });

            if (response.status !== 200) {
                console.log('BAD');
                this.installing = false;
                return;
            }

            $.popup({
                size: 'installer-popup',
                content: `
                    <div class="modal-header">
                        <button type="button" class="close" data-dismiss="modal" aria-hidden="true">&times;</button>
                        <h4 class="modal-title">Installing ${this.product.name}</h4>
                    </div>
                    <div class="modal-body"></div>
                    <div class="modal-footer">
                        <button type="button" class="btn btn-default" data-dismiss="modal">Blue Pill</button>
                        <button type="button" class="btn btn-primary" data-dismiss="modal">Red Pill</button>
                    </div>
                `
            });

            const popup = document.querySelector('.size-installer-popup');

            const chunk = async (response) => {
                const reader = response.body.getReader();
                const decoder = new TextDecoder();

                const parse = (str) => str.split("\n").filter((line) => !!line).map((line) => ({
                    type: line.substring(0, line.indexOf(': ')),
                    value: line.substring(line.indexOf(': ') + 2)
                }));

                const prepareMessage = (str) => {
                    ['INFO', 'ERROR'].forEach((status) => {
                        if (str.indexOf(status) === 0) {
                            str = `<span class="message-${status.toLowerCase()}">${status}</span> <pre>${str.substring(status.length + 1)}</pre>`;
                        }
                    });
                    return `<div class="install-message">${str}</div>`;
                };

                let done, value, message;
                while (!done) {
                    ({value, done} = await reader.read());
                    if (done) {
                        return;
                    }

                    message = parse(decoder.decode(value));

                    if (message[0].value === 'message') {
                        popup.querySelector('.modal-body').innerHTML += prepareMessage(message[1].value);
                    }

                    if (message[0].value === 'error') {
                        alert(message[1].value);
                    }
                }
            };

            await chunk(response);
            this.installing = false;
```

jaxwilko committed:
Report missing version.yaml as no update rather than failed update

jaxwilko committed:
Added fix for eslint

jaxwilko committed:
Added compiled js

jaxwilko committed:
Re-added old updates.js file

jaxwilko committed:
Added install support for plugins

LukeTowers commented on modules/system/controllers/updates/assets/src/components/Product.vue:
```diff
@@ -146,4 +215,37 @@ export default {
+    background-size: 50px 50px;
+    background-repeat: no-repeat;
+    background-position: 50% 50%;
+    background-image: url(/modules/system/assets/ui/images/loader-transparent.svg);
```
Any way for this to be loaded via the ASSET_URL?

jaxwilko replied:
yes there is... but this is all placeholder :innocent:

jaxwilko committed:
Added logic to remove composer plugin on uninstall

jaxwilko committed:
Added fix to better handle caching of local data

jaxwilko committed:
Improved ui

jaxwilko committed:
Added support for creating module asset configs

jaxwilko committed:
Added vite for updates/install page

jaxwilko committed:
Added missing files for module configging

jaxwilko committed:
Added ui fixes to install page

jaxwilko committed:
Added theme upload string

jaxwilko committed:
Manual develop merge

LukeTowers committed:
Merge branch 'wip/support-uploading-packages' into wip/support-uploading-packages-update-manager-changes

jaxwilko committed:
Manual merge

jaxwilko committed:
Added storm branch pick

LukeTowers committed:
Merge branch 'develop' into wip/support-uploading-packages-update-manager-changes

jaxwilko committed:
Added little fixes

jaxwilko committed:
Manual merge develop

jaxwilko committed:
Added wip and backend ui

jaxwilko committed:
Added fixes for marketplace container

jaxwilko committed:
Added missing manifest

jaxwilko committed:
Manual develop merge

jaxwilko commented:
Just pushed a big re-work of the backend UI. This push include [package-list.json](https://github.com/wintercms/winter/blob/ff8a4d7afdf1e42b8cffa230069097218f643129/modules/system/classes/core/package-list.json) which is a sample of demo data that will eventually be provided by the Winter Marketplace API.

As the sample is include and I've commit compiled assets for the backend, people should be able to test if they want.

> This is highly experimental! I've not had a chance to test the install functionality recently and as last time I did I managed to delete a load of work I would recommend testing in a fresh copy of Winter.

## Some Screenshots:
### Desktop
<img width="2116" height="1251" alt="image" src="https://github.com/user-attachments/assets/0b4551c9-7db4-408c-ae0c-29b84119199d" />
<img width="2116" height="1251" alt="image" src="https://github.com/user-attachments/assets/33077895-8c45-4160-8277-7254cc9099f1" />
<img width="2116" height="1251" alt="image" src="https://github.com/user-attachments/assets/de57fcdf-0417-488b-be79-53ce81167856" />
<img width="2116" height="1251" alt="image" src="https://github.com/user-attachments/assets/a1c642fb-4e54-4d0c-a5b1-ec4eb49d5267" />
<img width="1660" height="1253" alt="image" src="https://github.com/user-attachments/assets/0d3c4a70-42e1-4471-ae75-a50cec5a9e07" />

### Tablet
<img width="1256" height="1250" alt="image" src="https://github.com/user-attachments/assets/efde3b4d-dde9-4f83-8ca4-c3f9a2174306" />

### Mobile
<img width="566" height="1248" alt="image" src="https://github.com/user-attachments/assets/26486d81-1403-4fbb-a474-c72e5a772a6d" />

LukeTowers committed:
Move pagination below filters

Pagination was breaking the screen width on smaller screens and larger numbers of pages

jaxwilko committed:
Added fix for pagination overflowing page

jaxwilko committed:
Added support for passing in element size in pagination

jaxwilko committed:
Added fixes for updates popup

jaxwilko committed:
Added update indicator on product component

jaxwilko commented:
Pushed up some fixes for the update popup, need to test it actually works but it now loads which is a good first step.

Also added an update icon for products in the marketplace:
<img width="856" height="869" alt="image" src="https://github.com/user-attachments/assets/a249a856-562d-4958-8739-e8b8a8a5db9b" />
<img width="856" height="869" alt="image" src="https://github.com/user-attachments/assets/7ae61edf-3d6b-4c4b-81d8-13e056887126" />

jaxwilko committed:
Added fix for jittery popout

jaxwilko committed:
Added fixes to allow installing via the marketplace ui

jaxwilko committed:
Added fix for removing a plugin not rolling back the db

jaxwilko committed:
Added pagination to the bottom of marketplace as well as top

jaxwilko committed:
Added improved install ui and vue tippy for titles

jaxwilko committed:
Merged plugin list and plugin manage

jaxwilko committed:
Added some prettyness to the plugin manage page

jaxwilko committed:
Added tweaks to plugins update page

austinkregel commented:
@jaxwilko This looks really nice 🔥 

but it does appear that we're missing at least one interface 🤔

jaxwilko committed:
Added ui tweaks

jaxwilko commented:
Hey @austinkregel, which interface are you referring to?

jaxwilko commented:
Plugin Management page after some tweaks:
<img width="1682" height="1251" alt="image" src="https://github.com/user-attachments/assets/ed36ddec-ff04-46c3-9dd0-5d550931120e" />

jaxwilko committed:
Added support for pinning frozen plugins

jaxwilko committed:
Added wip to update interface

LukeTowers committed:
Replace usage of Composer::info() with Composer::show()

LukeTowers committed:
Merge branch 'wip/support-uploading-packages-update-manager-changes' of github.com:wintercms/winter into wip/support-uploading-packages-update-manager-changes

LukeTowers committed:
Make use of Composer::getLatestSupportedVersion()

LukeTowers committed:
Add todo

LukeTowers committed:
Use command.beforeRun event

LukeTowers committed:
Tidy imports

LukeTowers committed:
Comment

LukeTowers committed:
Style tweaks

LukeTowers committed:
Remove unused method

LukeTowers committed:
Tweaks

LukeTowers committed:
Default to only running migrations from the UpdateManager->update() method

LukeTowers committed:
Use faster composer methods from storm

LukeTowers committed:
Improve handling for removing plugins that don't exist

LukeTowers committed:
Fix usage of composer

LukeTowers committed:
Add comment

LukeTowers commented on composer.json:
```diff
@@ -81,7 +81,8 @@
             "composer/installers": true,
-            "wikimedia/composer-merge-plugin": true
+            "wikimedia/composer-merge-plugin": true,
+            "php-http/discovery": true
```
```suggestion
            "php-http/discovery": false
```

LukeTowers committed:
Update composer.json

LukeTowers committed:
Add support for getVersion to Theme

LukeTowers commented on composer.json:
```diff
@@ -30,7 +30,7 @@
     "require": {
         "php": ">=8.1",
-        "winter/storm": "dev-develop as 1.2",
+        "winter/storm": "dev-wip/support-uploading-packages-update-manager-changes as 1.2",
```
This will need to be updated before merge.

LukeTowers committed:
Code tidying

LukeTowers committed:
Tidy

LukeTowers committed:
Merge branch 'develop' into wip/support-uploading-packages-update-manager-changes

LukeTowers committed:
Lazy load core module console commands

LukeTowers committed:
Merge branch 'develop' into wip/support-uploading-packages-update-manager-changes

# Conflicts:
#	modules/backend/ServiceProvider.php

jaxwilko committed:
Added fix to ensure the right app is available at the right time

jaxwilko committed:
Added fix to ensure tests are compatible with newer versions of vite

jaxwilko committed:
Added fix for npm:version command description

LukeTowers committed:
Merge branch 'develop' into wip/support-uploading-packages-update-manager-changes

LukeTowers commented:
@coderabbitai review
